packages/effect/src/JSONSchema.ts

```diff
@@ -541,69 +556,82 @@ const go = (
-              enums.push(e)
+            const last = anyOf[anyOf.length - 1]
+            if (last !== undefined && isEnumOnly(last)) {
+              for (const e of schema.enum) {
+                last.enum.push(e)
+              }
+            } else {
+              anyOf.push(schema)
             }
           }
         } else {
           anyOf.push(schema)
         }
       }
-      if (anyOf.length === 0) {
-        return merge({ enum: enums }, getJsonSchemaAnnotations(ast))
+      if (anyOf.length === 1 && isEnumOnly(anyOf[0])) {
+        return { enum: anyOf[0].enum, ...getJsonSchemaAnnotations(ast) }
       } else {
-        if (enums.length >= 1) {
-          anyOf.push({ enum: enums })
-        }
-        return merge({ anyOf }, getJsonSchemaAnnotations(ast))
+        return { anyOf, ...getJsonSchemaAnnotations(ast) }
       }
     }
     case "Enums": {
-      return merge({
+      return {
         $comment: "/schemas/enums",
-        anyOf: ast.enums.map((e) => ({ title: e[0], enum: [e[1]] }))
-      }, getJsonSchemaAnnotations(ast))
+        anyOf: ast.enums.map((e) => ({ title: e[0], enum: [e[1]] })),
+        ...getJsonSchemaAnnotations(ast)
+      }
     }
     case "Refinement": {
-      if (AST.encodedBoundAST(ast) === ast) {
+      // The jsonSchema annotation is required only if the refinement does not have a transformation
+      if (AST.getTransformationFrom(ast) === undefined) {
         throw new Error(errors_.getJSONSchemaMissingAnnotationErrorMessage(path, ast))
       }
-      return go(ast.from, $defs, true, path)
+      return go(ast.from, $defs, handleIdentifier, path, options)
     }
     case "TemplateLiteral": {
       const regex = AST.getTemplateLiteralRegExp(ast)
-      return merge({
+      return {
         type: "string",
         title: String(ast),
         description: "a template literal",
-        pattern: regex.source
-      }, getJsonSchemaAnnotations(ast))
+        pattern: regex.source,
+        ...getJsonSchemaAnnotations(ast)
+      }
     }
     case "Suspend": {
       const identifier = Option.orElse(AST.getJSONIdentifier(ast), () => AST.getJSONIdentifier(ast.f()))
       if (Option.isNone(identifier)) {
         throw new Error(errors_.getJSONSchemaMissingIdentifierAnnotationErrorMessage(path, ast))
       }
-      return go(ast.f(), $defs, true, path)
+      return go(ast.f(), $defs, handleIdentifier, path, options)
     }
     case "Transformation": {
-      // Properly handle S.parseJson transformations by focusing on
-      // the 'to' side of the AST. This approach prevents the generation of useless schemas
-      // derived from the 'from' side (type: string), ensuring the output matches the intended
-      // complex schema type.
       if (isParseJsonTransformation(ast.from)) {
-        return go(ast.to, $defs, true, path)
+        switch (options.parseJsonStrategy) {
+          case "to":
+            return go(ast.to, $defs, handleIdentifier, path, options)
```

Comment:
Hmm wouldn't this generate invalid schemas? If a schema is expecting a json "string" this would result in a parse error.

Reply:
@tim-smart the special case was introduced to address [https://github.com/Effect-TS/effect/issues/3086](https://github.com/Effect-TS/effect/issues/3086), but I suppose this behavior shouldn't be recursive as it is now. It should only apply once at the top level